libp2p/protocols/kademlia/provider.nim

```diff
@@ -7,12 +7,53 @@
 # This file may not be copied, modified, or distributed except according to
 # those terms.
 
-import std/sequtils
+import std/[sequtils, tables, sets, heapqueue]
 import chronos, chronicles, results
 import ../../[peerid, switch, multihash, cid]
+import ../../utils/heartbeat
 import ../protocol
 import ./[protobuf, types, find]
 
+proc `==`*(a, b: ProviderRecord): bool =
+  a.provider.id == b.provider.id and a.cid == b.cid
+
+# for HeapQueue
+proc `<`*(a, b: ProviderRecord): bool =
+  a.expiresAt < b.expiresAt
+
+proc `<`*(a: ProviderRecord, b: chronos.Moment): bool =
+  a.expiresAt < b
+
+proc addProviderRecord(pm: ProviderManager, record: ProviderRecord) =
+  if not pm.knownCids.hasKey(record.cid):
+    pm.knownCids[record.cid] = initHashSet[Provider]()
+
+  try:
+    pm.knownCids[record.cid].incl(record.provider)
+
+    # remove old providerRecord if any
+    let oldRecordIdx = pm.records.find(record)
+    if oldRecordIdx != -1:
+      pm.records.del(oldRecordIdx)
+
+    # push new providerRecord
+    pm.records.push(record)
+  except KeyError:
+    raiseAssert("checked with hasKey")
+
+proc removeProviderRecord(pm: ProviderManager, record: ProviderRecord) =
+  ## Remove provider record and related keys
+
+  if not pm.knownCids.hasKey(record.cid):
+    return
+
+  try:
+    pm.knownCids[record.cid].excl(record.provider)
+    if pm.knownCids[record.cid].len() == 0:
+      pm.knownCids.del(record.cid)
+  except KeyError:
+    raiseAssert("checked with hasKey")
```

Comment:
```nim
  try:
    pm.knownCids[record.cid].excl(record.provider)
    if pm.knownCids[record.cid].len() == 0:
      pm.knownCids.del(record.cid)
  except KeyError:
    discard # do nothing when knownCids don't have cid
```

consider that code don't really need `if not pm.knownCids.hasKey(record.cid): return`.  
this is not suggestion, just fyi.